nearcore/src/cold_storage.rs

```diff
@@ -0,0 +1,189 @@
+use std::sync::{atomic::AtomicBool, Arc};
+
+use near_chain::types::Tip;
+use near_epoch_manager::EpochManagerAdapter;
+use near_primitives::{hash::CryptoHash, types::BlockHeight};
+use near_store::{
+    cold_storage::{update_cold_db, update_cold_head},
+    db::ColdDB,
+    DBCol, NodeStorage, Store, FINAL_HEAD_KEY, HEAD_KEY,
+};
+
+use crate::{NearConfig, NightshadeRuntime};
+
+/// A handle that keeps the state of the cold store loop and can be used to stop it.
+pub struct ColdStoreLoopHandle {
+    join_handle: std::thread::JoinHandle<()>,
+    keep_going: Arc<AtomicBool>,
+}
+
+impl ColdStoreLoopHandle {
+    pub fn stop(self) {
+        self.keep_going.store(false, std::sync::atomic::Ordering::SeqCst);
+        match self.join_handle.join() {
+            Ok(_) => {
+                tracing::debug!(target:"cold_store", "Joined the cold store loop thread");
+            }
+            Err(_) => {
+                tracing::error!(target:"cold_store", "Failed to join the cold store loop thread");
+            }
+        }
+    }
+}
+
+/// The status of the cold store loop indicates if the loop should sleep before
+/// the next copy attempt or if it should actively continue copying.
+enum ColdStoreLoopStatus {
+    Sleep,
+    Continue,
```

Comment:
nit: 

rather than sleep/continue, I'd rename this enum to 'ColdStoreCopyResult' - and have values like 'NoBlocksCopied' or 'BlockCopied' etc. 

this way you have tiny bit nicer separation (as based on this - you can 'later' decide whether you sleep or not)